site/docs/v1/tech/migration-guide/duende.adoc

```diff
@@ -0,0 +1,331 @@
+
+* `date`
+* `json`
+* `optparse`
+* `securerandom`
+* `fusionauth_client`
+
+It is likely that all of these will be on your system already, except the `fusionauth_client` gem.
+
+If you have `bundler` installed, run `bundle install` in the `duende` directory. Otherwise, install the needed gems another way.
+
+=== Use the Script
+
+You can see the output of the script by running it with the `-h` option:
+
+
+
+[source,sh,title=Running the import script with the help command line switch]
+----
+ruby ./import.rb -h
+----
+
+The output will be similar to this:
+
+[source,sh,title=The help output of the import.rb script]
+----
+Usage: import.rb [options]
+    -l, --link-social-accounts       Link social accounts, if present, after import. This operation is slower than an import.
+    -r APPLICATION_IDS,              A comma separated list of existing applications Ids. All users will be registered for these applications.
+        --register-users
+    -o, --only-link-social-accounts  Link social accounts with no import.
+    -u, --users-file USERS_FILE      The exported JSON user data file from IdentityServer. Defaults to users.json.
+    -f FUSIONAUTH_URL,               The location of the FusionAuth instance. Defaults to http://localhost:9011.
+        --fusionauth-url
+    -k, --fusionauth-api-key API_KEY The FusionAuth API key.
+    -t TENANT_ID,                    The FusionAuth tenant id. Required if more than one tenant exists.
+        --fusionauth-tenant-id
+    -h, --help                       Prints this help.
+----
+
+For this script to work correctly, set the following switches, unless the defaults work for you:
+
+* `-u` should point to the location of the user export file you obtained, unless the default works.
+* `-f` should point to your FusionAuth instance. If you are testing locally, it will probably be `\http://localhost:9011`.
+* `-k` should be set to the value of the API key created above.
+* `-t` should be set to the Id of the testing tenant created above.
+
+The `-o` and `-l` switches will attempt to create links for any users authenticated via Google or another social identity provider found in the users data file.
+
+If you are loading users with social account authentication, you must create the social identity providers in FusionAuth beforehand or the links will fail. Additionally, creating a link is not currently optimized in the same way that loading a user is. It may make sense to import all the users in one pass (omitting the `-l` switch) and then create the links using the `-o` switch in a second pass, after the users are imported.
+
+[NOTE.note]
+====
+The social account linking functionality will only work with FusionAuth version 1.28 or above. The `fusionauth_client` library must be >= 1.28.
+====
+
+When you run the script, you should get an output similar to the following:
+
+[source,shell,title=Import script output]
+----
+$ ruby ./import.rb -f http://localhost:9011 -k '...' -u user-data.json
```

Comment:
Could you add `-t '...'` to the command. I missed this on my first try, which caused the import.rb to crash with a non-descriptive error message:
```
Traceback (most recent call last):
	2: from ./import.rb:240:in `<main>'
	1: from ./import.rb:156:in `import'
./import.rb:156:in `+': no implicit conversion of nil into String (TypeError)
```